runtime/js/40_command.js

```diff
@@ -0,0 +1,214 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+"use strict";
+
+((window) => {
+  const core = window.Deno.core;
+  const { pathFromURL } = window.__bootstrap.util;
+  const { read, write } = window.__bootstrap.io;
+  const { illegalConstructorKey } = window.__bootstrap.webUtil;
+  const { ArrayPrototypeMap, ObjectEntries, String, TypeError } =
+    window.__bootstrap.primordials;
+
+  function createWritableIOStream(rid) {
+    return new WritableStream({
+      async write(chunk) {
+        let nwritten = 0;
+        while (nwritten < chunk.byteLength) {
+          nwritten += await write(rid, chunk.subarray(nwritten));
+        }
+      },
+      abort() {
+        core.tryClose(rid);
+      },
+    });
+  }
+
+  function createReadableIOStream(rid) {
+    return new ReadableStream({
+      async pull(controller) {
+        const view = controller.byobRequest.view;
+        const res = await read(rid, view);
+        if (res === null) {
+          core.close(rid);
+          controller.close();
+        } else {
+          controller.byobRequest.respond(res);
+        }
+      },
+      cancel() {
+        core.close(rid);
+      },
+      type: "bytes",
+      autoAllocateChunkSize: 16384,
+    });
+  }
+
+  class Command {
+    #rid;
+
+    constructor(command, {
+      args = [],
+      cwd = undefined,
+      clearEnv = false,
+      env = {},
+      uid = undefined,
+      gid = undefined,
+    } = {}) {
+      this.#rid = core.opSync("op_create_command", {
+        cmd: pathFromURL(command),
+        args: ArrayPrototypeMap(args, String),
+        cwd: pathFromURL(cwd),
+        clearEnv,
+        env: ObjectEntries(env),
+        uid,
+        gid,
+      });
+    }
+
+    spawn(options = {}) {
+      if (this.#rid === null) {
+        throw new TypeError("Child has already been used up.");
+      }
+      const child = core.opSync("op_command_spawn", this.#rid, options);
+      this.#rid = null;
+      return new Child(illegalConstructorKey, child);
+    }
+
+    async status(options = {}) {
+      if (this.#rid === null) {
+        throw new TypeError("Child has already been used up.");
+      }
+      const status = await core.opAsync(
+        "op_command_status",
+        this.#rid,
+        options,
+      );
+      this.#rid = null;
```

Comment:
What happens when I call `p.status(); p.status();`? (I.e., without waiting for the first promise to resolve.)

My guess is it calls op_command_status twice and that's probably bad. I'd write this as:
```suggestion
      const rid = this.#rid;
      this.#rid = null;
      const status = await core.opAsync(
        "op_command_status",
        rid,
        options,
      );
```
Also applies to wait() and output() (both methods.)

(Maybe calling it twice here is benign and the real bug is clearing `this.#rid` but the general advice still applies to the other methods.)

Reply:
I can do this for the `Command` class, but this becomes a bit problematic with the `Child` class, due to the `status` getter